src/main/java/de/tum/in/www1/artemis/service/tutorialgroups/TutorialGroupScheduleService.java

```diff
@@ -0,0 +1,172 @@
+import de.tum.in.www1.artemis.domain.tutorialgroups.TutorialGroup;
+import de.tum.in.www1.artemis.domain.tutorialgroups.TutorialGroupSchedule;
+import de.tum.in.www1.artemis.domain.tutorialgroups.TutorialGroupSession;
+import de.tum.in.www1.artemis.domain.tutorialgroups.TutorialGroupsConfiguration;
+import de.tum.in.www1.artemis.repository.tutorialgroups.TutorialGroupScheduleRepository;
+import de.tum.in.www1.artemis.repository.tutorialgroups.TutorialGroupSessionRepository;
+import de.tum.in.www1.artemis.web.rest.tutorialgroups.TutorialGroupDateUtil;
+import de.tum.in.www1.artemis.web.rest.tutorialgroups.errors.ScheduleOverlapsWithSessionException;
+
+@Service
+public class TutorialGroupScheduleService {
+
+    private final TutorialGroupSessionRepository tutorialGroupSessionRepository;
+
+    private final TutorialGroupScheduleRepository tutorialGroupScheduleRepository;
+
+    private final TutorialGroupFreePeriodService tutorialGroupFreePeriodService;
+
+    public TutorialGroupScheduleService(TutorialGroupSessionRepository tutorialGroupSessionRepository, TutorialGroupScheduleRepository tutorialGroupScheduleRepository,
+            TutorialGroupFreePeriodService tutorialGroupFreePeriodService) {
+        this.tutorialGroupSessionRepository = tutorialGroupSessionRepository;
+        this.tutorialGroupScheduleRepository = tutorialGroupScheduleRepository;
+        this.tutorialGroupFreePeriodService = tutorialGroupFreePeriodService;
+    }
+
+    /**
+     * Create a new schedule for the given tutorial group and creates all corresponding sessions
+     *
+     * @param tutorialGroupsConfiguration the course wide tutorial groups configuration
+     * @param tutorialGroup               the tutorial group for which the schedule should be created
+     * @param tutorialGroupSchedule       the schedule to create
+     */
+    public void saveScheduleAndGenerateScheduledSessions(TutorialGroupsConfiguration tutorialGroupsConfiguration, TutorialGroup tutorialGroup,
+            TutorialGroupSchedule tutorialGroupSchedule) {
+        // Generate Sessions
+        var individualSessions = generateSessions(tutorialGroupsConfiguration, tutorialGroupSchedule);
+        // check for overlap with existing individual sessions
+        var overlappingIndividualSessions = new HashSet<TutorialGroupSession>();
+        for (var individualSession : individualSessions) {
+            var overlappingSession = tutorialGroupSessionRepository.findOverlappingIndividualSessionsInSameTutorialGroup(tutorialGroup, individualSession.getStart(),
+                    individualSession.getEnd());
+            if (!overlappingSession.isEmpty()) {
+                overlappingIndividualSessions.addAll(overlappingSession);
+            }
+        }
+        if (!overlappingIndividualSessions.isEmpty()) {
+            throw new ScheduleOverlapsWithSessionException(overlappingIndividualSessions, ZoneId.of(tutorialGroupsConfiguration.getCourse().getTimeZone()));
+        }
```

Comment:
Consider moving this into a method.

Reply:
Good idea! Done in 0a94757fc35450f1ba895a7ac0260d3bd3d8515b